Design.md

```diff
@@ -0,0 +1,149 @@
+
+Authentication is implemented with mTLS. Server and client have different CA roots, which they are expected to exchange in a secure way. (Scenario-wise: the client CA is operated by the same organization that hosts job-runner). The server identifies each client with it's CommonName (CN) field of the certificate. The client CA only issues certificates with CN values to developers that are allowed to access the API. All running jobs are bound to the CN of the client calling `Start`, and only a client with that certificate is allowed to execute operations for that job.
+
+A more complete system should also support certificate revocation lists (CRL), so that individual certificates can be revoked in case they are compromised. Other useful feature would be having more fine-grained access control per certificate, for instance restricting clients to only specific endpoints, executables, and resource limits.
+
+## Endpoints
+
+### Start
+
+Starts a new job by spawning a process from given executable path and arguments. Returns a unique job id (UUID v4), that is used to specify the target job for other endpoints. If the executable is not found or cannot be executed, immediately returns an error.
+
+No security checks are applied to the program and arguments. However, as the job is placed inside an isolated container, it shouldn't be able to do much damage. It can still consume resources in the limits that cgroup-limits allow, and it can flood it's output with thrash data.
+
+### Stop
+
+Cancels a job by sending `SIGKILL`. Stop only returns after the process has been terminated, and returns the status of the job. If job was already terminated before calling stop, the old state is returned without attempting termination.
+
+Could be improved by sending `SIGTERM` shortly before `SIGKILL`, but I'm aiming for simplicity here. Also separating soft and hard kills in the API level might be useful in some situations.
+
+### Status
+
+Returns job status, i.e. is it running, and the status code if the job has completed. If the job has been terminated with a signal, that is reported instead.
+
+```proto
+message JobStatus {
+    enum Status {
+        RUNNING = 0;    // Still running (value is zero)
+        COMPLETED = 1;  // Completed normally (value is return code)
+        SIGNAL = 2;     // Terminated by a signal (value is signal number)
+        MISSING = 3;    // No such process (or access denied) (value is zero)
+    }
+    Status status = 1;
+    int32 value = 2;
```

Comment:
What is value?

Reply:
Explained by the comments in the above enum.

Bike-shedding:
* `return_code` might be good otherwise, but then signal would have a wrong name. 
* `status` is already used by the field above.
* Using `oneof` instead of `enum` + field would be preferred if protobuf supported proper sum-types, but it doesn't and `oneof` with empty/dummy fields seems ugly as well.